File: backend/src/models/transaction.rs
// Copyright (c) 2024 Tempo Budget
// SPDX-License-Identifier: MIT
//
// Transaction-related data models.

//! # Transaction Models
//!
//! This module defines transaction-related data structures including regular
//! transactions and recurring transaction templates.

use serde::{Deserialize, Serialize};
use sqlx::FromRow;

/// Represents a financial transaction (income or expense).
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct Transaction {
    /// Unique identifier (UUID)
    pub id: String,
    /// ID of the budget this transaction belongs to
    pub budget_id: String,
    /// ID of the category this transaction is assigned to
    pub category_id: String,
    /// Transaction title/description
    pub title: String,
    /// Transaction amount (positive value)
    pub amount: f64,
    /// Type of transaction: "income" or "expense"
    pub transaction_type: String,
    /// Date of the transaction (ISO 8601 format)
    pub date: String,
    /// Optional comment or note
    pub comment: Option<String>,
    /// Whether this transaction was generated from a recurring template (0=no, 1=yes)
    pub is_recurring: i32,
    /// Timestamp when the transaction was created (RFC 3339 format)
    pub created_at: String,
}

/// Request payload for creating a new transaction.
#[derive(Debug, Deserialize)]
pub struct CreateTransaction {
    /// ID of the budget for the transaction
    pub budget_id: String,
    /// ID of the category for the transaction
    pub category_id: String,
    /// Transaction title/description
    pub title: String,
    /// Transaction amount
    pub amount: f64,
    /// Type of transaction: "income" or "expense"
    pub transaction_type: String,
    /// Date of the transaction (ISO 8601 format)
    pub date: String,
    /// Optional comment or note
    pub comment: Option<String>,
}

/// Represents a recurring transaction template.
///
/// Recurring transactions are templates that can automatically generate
/// regular transactions based on a specified frequency.
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct RecurringTransaction {
    /// Unique identifier (UUID)
    pub id: String,
    /// ID of the budget this recurring transaction belongs to
    pub budget_id: String,
    /// ID of the category for generated transactions
    pub category_id: String,
    /// Transaction title/description
    pub title: String,
    /// Transaction amount
    pub amount: f64,
    /// Type of transaction: "income" or "expense"
    pub transaction_type: String,
    /// Frequency: "daily", "weekly", "monthly", "yearly"
    pub frequency: String,
    /// Day of the period (e.g., day of month for monthly frequency)
    pub day: Option<i32>,
    /// Whether the recurring transaction is active (0=inactive, 1=active)
    pub active: i32,
    /// Timestamp when the recurring transaction was created (RFC 3339 format)
    pub created_at: String,
}

/// Request payload for creating a new recurring transaction.
#[derive(Debug, Deserialize)]
pub struct CreateRecurringTransaction {
    /// ID of the budget for the recurring transaction
    pub budget_id: String,
    /// ID of the category for generated transactions
    pub category_id: String,
    /// Transaction title/description
    pub title: String,
    /// Transaction amount
    pub amount: f64,
    /// Type of transaction: "income" or "expense"
    pub transaction_type: String,
    /// Frequency: "daily", "weekly", "monthly", "yearly"
    pub frequency: String,
    /// Day of the period (optional, depends on frequency)
    pub day: Option<i32>,
}

File: backend/app/models/transaction.py
# Copyright (c) 2024 Tempo Budget
# SPDX-License-Identifier: MIT
"""Transaction-related data models."""

from typing import Literal
from pydantic import BaseModel, Field


class Transaction(BaseModel):
    """Represents a financial transaction."""
    id: str = Field(..., description="Unique identifier (UUID)")
    budget_id: str = Field(..., description="ID of the budget")
    category_id: str = Field(..., description="ID of the category")
    title: str = Field(..., description="Transaction title/description")
    amount: float = Field(..., description="Transaction amount")
    transaction_type: str = Field(..., description="Type: 'income' or 'expense'")
    date: str = Field(..., description="Date of the transaction")
    comment: str | None = Field(None, description="Optional comment")
    is_recurring: int = Field(..., description="Whether from recurring template (0/1)")
    paid_by_user_id: str | None = Field(None, description="ID of user who paid (for group budgets)")
    created_at: str = Field(..., description="Creation timestamp")

    class Config:
        from_attributes = True


class CreateTransaction(BaseModel):
    """Request payload for creating a new transaction."""
    category_id: str = Field(..., description="ID of the category")
    title: str = Field(..., min_length=1, description="Transaction title")
    amount: float = Field(..., gt=0, description="Transaction amount")
    transaction_type: Literal["income", "expense"] = Field(..., description="Type")
    date: str = Field(..., description="Date (ISO 8601 format)")
    comment: str | None = Field(None, description="Optional comment")
    paid_by_user_id: str | None = Field(None, description="ID of user who paid (for group budgets)")


class RecurringTransaction(BaseModel):
    """Represents a recurring transaction template."""
    id: str = Field(..., description="Unique identifier (UUID)")
    budget_id: str = Field(..., description="ID of the budget")
    category_id: str = Field(..., description="ID of the category")
    title: str = Field(..., description="Transaction title")
    amount: float = Field(..., description="Transaction amount")
    transaction_type: str = Field(..., description="Type: 'income' or 'expense'")
    frequency: str = Field(..., description="Frequency: daily/weekly/monthly/yearly")
    day: int | None = Field(None, description="Day of the period")
    active: int = Field(..., description="Whether active (0/1)")
    created_at: str = Field(..., description="Creation timestamp")

    class Config:
        from_attributes = True


class CreateRecurringTransaction(BaseModel):
    """Request payload for creating a recurring transaction."""
    category_id: str = Field(..., description="ID of the category")
    title: str = Field(..., min_length=1, description="Transaction title")
    amount: float = Field(..., gt=0, description="Transaction amount")
    transaction_type: Literal["income", "expense"] = Field(..., description="Type")
    frequency: Literal["daily", "weekly", "monthly", "yearly"] = Field(..., description="Frequency")
    day: int | None = Field(None, description="Day of the period")
